pkg/apis/allocation/v1/gameserverallocation.go

```diff
@@ -278,6 +292,33 @@ func (gsa *GameServerAllocation) Validate() ([]metav1.StatusCause, bool) {
 			causes = append(causes, c...)
 		}
 	}
+	for i := range gsa.Spec.Selectors {
+		if c, ok := gsa.Spec.Selectors[i].Validate(fmt.Sprintf("spec.selectors[%d]", i)); !ok {
+			causes = append(causes, c...)
+		}
+	}
+
+	if (len(gsa.Spec.Required.MatchLabels) != 0 || len(gsa.Spec.Required.MatchExpressions) != 0 || gsa.Spec.Required.GameServerState != nil || gsa.Spec.Required.Players != nil || len(gsa.Spec.Preferred) != 0) && len(gsa.Spec.Selectors) != 0 {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueDuplicate,
+			Message: "Selectors field can only be set if neither Required or Preferred is set.",
+			Field:   "spec.selectors",
+		})
+	}
 
 	return causes, len(causes) == 0
 }
+
+// Converter converts game server allocation required and preferred fields to selectors field.
+func (gsa *GameServerAllocation) Converter() {
+	if len(gsa.Spec.Selectors) == 0 {
+		var selectors []GameServerSelector
+		selectors = append(selectors, gsa.Spec.Preferred...)
+		if len(gsa.Spec.Required.MatchLabels) != 0 || len(gsa.Spec.Required.MatchExpressions) != 0 || gsa.Spec.Required.GameServerState != nil || gsa.Spec.Required.Players != nil {
```

Comment:
🤔 do we have to check all the fields of `gsa.Spec.Required`? - If it's empty, I'm fairly sure we could add it - it resolves to allowing everything which should be fine (also means that if we add more fields, we don' t have to remember to come back and fix things here).

Reply:
Done.